src/librustc/ty/maps/job.rs

```diff
@@ -93,5 +147,379 @@ impl<'tcx> QueryJob<'tcx> {
+    fn new() -> Self {
+        QueryLatch {
+            info: Mutex::new(QueryLatchInfo {
+                complete: false,
+                waiters: Vec::new(),
+            }),
+        }
+    }
+
+    /// Awaits the caller on this latch by blocking the current thread.
+    fn await(&self, waiter: &Lrc<QueryWaiter<'tcx>>) {
+        let mut info = self.info.lock();
+        if !info.complete {
+            // We push the waiter on to the `waiters` list. It can be accessed inside
+            // the `wait` call below, by 1) the `set` method or 2) by deadlock detection.
+            // Both of these will remove it from the `waiters` list before resuming
+            // this thread.
+            info.waiters.push(waiter.clone());
+
+            // If this detects a deadlock and the deadlock handler wants to resume this thread
+            // we have to be in the `wait` call. This is ensured by the deadlock handler
+            // getting the self.info lock.
+            rayon_core::mark_blocked();
+            waiter.condvar.wait(&mut info);
+        }
+    }
+
+    /// Sets the latch and resumes all waiters on it
+    fn set(&self) {
+        let mut info = self.info.lock();
+        debug_assert!(!info.complete);
+        info.complete = true;
+        let registry = rayon_core::Registry::current();
+        for waiter in info.waiters.drain(..) {
+            waiter.notify(&registry);
+        }
+    }
+
+    /// Remove a single waiter from the list of waiters.
+    /// This is used to break query cycles.
+    fn extract_waiter(
+        &self,
+        waiter: usize,
+    ) -> Lrc<QueryWaiter<'tcx>> {
+        let mut info = self.info.lock();
+        debug_assert!(!info.complete);
+        // Remove the waiter from the list of waiters
+        info.waiters.remove(waiter)
+    }
+}
+
+/// A pointer to an active query job. This is used to give query jobs an identity.
+#[cfg(parallel_queries)]
+type Ref<'tcx> = *const QueryJob<'tcx>;
+
+/// A resumable waiter of a query. The usize is the index into waiters in the query's latch
+#[cfg(parallel_queries)]
+type Waiter<'tcx> = (Ref<'tcx>, usize);
+
+/// Visits all the non-resumable and resumable waiters of a query.
+/// Only waiters in a query are visited.
+/// `visit` is called for every waiter and is passed a query waiting on `query_ref`
+/// and a span indicating the reason the query waited on `query_ref`.
+/// If `visit` returns Some, this function returns.
+/// For visits of non-resumable waiters it returns the return value of `visit`.
+/// For visits of resumable waiters it returns Some(Some(Waiter)) which has the
+/// required information to resume the waiter.
+/// If all `visit` calls returns None, this function also returns None.
+#[cfg(parallel_queries)]
+fn visit_waiters<'tcx, F>(query_ref: Ref<'tcx>, mut visit: F) -> Option<Option<Waiter<'tcx>>>
+where
+    F: FnMut(Span, Ref<'tcx>) -> Option<Option<Waiter<'tcx>>>
+{
+    let query = unsafe { &*query_ref };
```

Comment:
This is a safe `fn` and the first thing it does is dereference a raw pointer. `ref` is a type alias, not a newtype, so it is trivial to safely create it from any kind of bad pointer value. Shouldn't the function be `unsafe` and document that it relies on the caller to uphold certain invariants?

(I have no idea how this kind of stuff is usually handled in the compiler, but this seems like an obvious candidate for an unsafe function to me. It doesn't matter that it's private, users within this module could still need a reminder to be careful.)

Reply:
I agree with your general assessment that the function ought to be declared `unsafe`, and have its contract carefully documented (that `Ref` is a valid job?). The fact that this makes the body "one big unsafe block" is a sort of unfortunate thing but not that big a deal (imo). 

That said, perhaps even better would be to change the signature to:

```rust
fn visit_waiters<'tcx, F>(query_ref: &QueryJob,'tcx>, ...)
```

and thus have the *caller* be the one who asserts safety. I've not read the code deeply enough yet to say whether this would be an improvement.

Reply:
I've removed the use of raw pointers to access queries and instead use `Lrc`.

Reply:
Awesome, thanks a lot. :)